docs/madr/decisions/049-meshexternalservice-api.md

```diff
@@ -0,0 +1,474 @@
+    * **value**: depends of the type can be an existing domain, new domain name, CIDR or IP
+    * **port**: defines a port to which a user does requests
+    * **protocol**: defines a protocol of the communication. Possible values:
+      * `tls`: should be used when TLS traffic is originated by the client application
+      * `tcp`: WARNING: shouldn't be used when match has only domains. On the TCP level we are not able to disinguish domain, in this case it is going to hijack whole traffic on this port.
+      * `grpc`
+      * `http`
+      * `http2`
+  * **destination**: defines where matched requests should be routed
+    * **type**: defines what kind of destination is it, one of `Regular`, `Passthrough`, or `Extension`, (Default: `Passthrough`)
+      * `Regular`: allows creating a set of destination endpoints and `TLS` configuration, when defined section `extension` is not available.
+      * `Passthrough`: traffic just passes a proxy without any modifications to the original destination, when defined sections `endpoints`, `tls` and `extension` are not available.
+      * `Extension`: allows specifying a custom plugin for example, user can create a plugin which support AWS Lambda, when defined sections `endpoints` and `tls` are not available.
+    * **extension**: struct for a plugin configuration
+      * **type**: defines what kind of plugin to use, it's a string type so any new plugins should works.
+      * **config**: json map that is mapped to configuration provided in the type.
+    * **endpoints**: defines a list of endpoints
+      * **address**: defines an address to which a user want to send a request. Is possible to provide `domain`, `ip` and `unix` sockets
+      * **port**: defines a port of a destination.
+    * **tls**: provides a TLS configuration when proxy is resposible for a TLS origination
+      * **enabled**: defines if proxy should originate TLS.
+      * **version**: section for providing version specification.
+        * **min**: defines minmum supported version. One of `TLS_AUTO`, `TLSv1_0`, `TLSv1_1`, `TLSv1_2`, `TLSv1_3`
+        * **max**: defines maximum supported version. One of `TLS_AUTO`, `TLSv1_0`, `TLSv1_1`, `TLSv1_2`, `TLSv1_3`
+      * **allowRenegotiation**: defines if TLS sessions will allow renegotiation.
+      * **verification**: section for providing TLS verification details.
+        * **skip**: defines if proxy should skip SAN verification. Default `false`.
+        * **subjectAltNames**: list of names to verify in the certificate.
+      * **caCert**: defines a certificate of CA.
+        * one of `inline`, `inlineString` or `secret`.
+      * **clientCert**: defines a certificate of a client.
+        * one of `inline`, `inlineString` or `secret`.
+      * **clientKey**: defines a client private key.
+        * one of `inline`, `inlineString` or `secret`.
+* **status**: status of an object managed by Kuma control-plane
+  * **vips**: list of allocated VIPs
```

Comment:
shouldn't this be
```
status:
  vip:
    type: Kuma
    ip: 240.0.0.2
  hostnames:
  - ...
```
?

Why separate hostnames needs separate VIPs?

I think we also need to have status of hostnames for conflicts?

Reply:
I initially thought we might require separate IP addresses for TCP traffic, but it seems unnecessary as we can match addresses and ports using a filter chain.

Reply:
WDYT about a section like: 
```
  addresses:
  - address: example.ext.svc.local
    status: Available
  - address: httpbin.com
    status: Available
  - address: 192.168.0.1
    status: Available
  - address: 10.1.1.0/24
    status: NotAvailable
    reason: "addresses are overlapping with ext2"
```

It makes sense to have domains and IPs, in case of IP overlapping, we can provide reason

Reply:
Is it even realistic to detect collisions? It's easy to compare IPs, but hostnames vs hostnames with wildcards? IP vs CIDRs?

Reply:
I think IPs vs CIDR can achieve through libraries [lib example](https://github.com/seancfoley/ipaddress-go/wiki/Code-Examples-2:-Subnet-Containment,-Matching,-Comparing#select-cidr-prefix-block-subnets-intersecting-with-arbitrary-block), hostnames are more tricky, simple thing we could do is `ends with`, but at the beginning, we might introduce only some simple case